### PR TITLE
[MME] Apn correction support

### LINKED ISSUE #2643: Add APN Correction
[please attach design doc / any description]

### LINKED ISSUE #2650: [APN Correction]
- APN Correction: Virtual APN
  - Send different APN on PDPContextCreation based on user profile (ex: charging characristics: prepaid, postpaid)
- APN Correction: Based on IMSI prefix
  - Send different APN on PDPContextCreation based IMSI prefix

### PR DESCRIPTION
**Pull request by ymasmoudi**
## Summary

Support APN Correction in MME
Add 2 integration tests to cover PDN connectivity request and ESM Information response.

## Test Plan

Run Unit tests and LTE Integration tests

**make integ_test TESTS=s1aptests/test_attach_esm_information_apn_correction.py** 
. /home/vagrant/build/python/bin/activate
echo "Running test: s1aptests/test_attach_esm_information_apn_correction.py"
Running test: s1aptests/test_attach_esm_information_apn_correction.py
sudo -E PATH=/usr/lib/ccache:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin PYTHONPATH=:/home/vagrant/s1ap-tester/bin /home/vagrant/build/python/bin/nosetests -x -s s1aptests/test_attach_esm_information_apn_correction.py || exit 1
tag: AUTH_TYPE tagVal: MILENAGE
************************* Enb tester config

**********************TPT open server******** 0

 Server Socket FD =[12]
************* Testing *Time elapsed(in usec): for timer expiry               :8382 usec
AGW is stateful
APN Correction configured
************************* restarting mme
....
Waiting mme restart for 14 seconds
Using subscriber IMSI IMSI001010000000001
************************* UE device config for ue_id  1
************************* UE device config for ue_id  2
************************* Configuring IP block
************************* Waiting for IP changes to propagate
************************* S1 setup
************************* UE App config
************************* sending Attach Request for ue-id : 1
Sending Attach Request ue-id 1
Received auth req ind 
Sending Auth Response ue-id 1
Received Security Mode Command ue-id 1
Received Esm Information Request ue-id 1
Sending Esm Information Response ue-id 1
Sending Esm Information with wrong APN  oai
Building ESM Information Response for Ue
PDN_TYPE: 1
PDN_ADDRESS: 192 168 128 240

**make integ_test TESTS=s1aptests/test_attach_esm_information_apn_correction.py** 
. /home/vagrant/build/python/bin/activate
echo "Running test: s1aptests/test_attach_esm_information_apn_correction.py"
Running test: s1aptests/test_attach_esm_information_apn_correction.py
sudo -E PATH=/usr/lib/ccache:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin PYTHONPATH=:/home/vagrant/s1ap-tester/bin /home/vagrant/build/python/bin/nosetests -x -s s1aptests/test_attach_esm_information_apn_correction.py || exit 1
tag: AUTH_TYPE tagVal: MILENAGE
************************* Enb tester config

**********************TPT open server  0

 Server Socket FD =[12]
************* Testing *Time elapsed(in usec): for timer expiry               :8715 usec
AGW is stateful
APN Correction configured
************************* restarting mme
....
Waiting mme restart for 14 seconds
Using subscriber IMSI IMSI001010000000001
************************* UE device config for ue_id  1
************************* UE device config for ue_id  2
************************* Configuring IP block
************************* Waiting for IP changes to propagate
************************* S1 setup
************************* UE App config
************************* sending Attach Request for ue-id : 1
Sending Attach Request ue-id 1
Received auth req ind 
Sending Auth Response ue-id 1
Received Security Mode Command ue-id 1
Received Esm Information Request ue-id 1
Sending Esm Information Response ue-id 1
Sending Esm Information with wrong APN  oai
Building ESM Information Response for Ue
PDN_TYPE: 1
PDN_ADDRESS: 192 168 128 76

**Unit Tests**

ninja -C   /home/vagrant/build/c//oai
ninja: Entering directory `/home/vagrant/build/c//oai'
[12/12] Linking CXX executable oai_mme/mme
cd  /home/vagrant/build/c//oai && ctest --output-on-failure
Test project /home/vagrant/build/c/oai
    Start 1: test_mme_app_ue_context
1/4 Test #1: test_mme_app_ue_context ..........   Passed    0.04 sec
    Start 2: test_openflow_controller
2/4 Test #2: test_openflow_controller .........   Passed    0.03 sec
    Start 3: test_imsi_encoder
3/4 Test #3: test_imsi_encoder ................   Passed    0.02 sec
    Start 4: test_gtp_app
4/4 Test #4: test_gtp_app .....................   Passed    0.03 sec

100% tests passed, 0 tests failed out of 4

### CODE REVIEW

File: lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py [hunk 1]
```diff
@@ -525,6 +526,31 @@ def update_mme_config_for_sanity(self, cmd):
...
 
+    def config_apn_correction(self, cmd):
+        """
+        Configure the apn correction mode on the access gateway
+
+        Args:
+          cmd: Specify how to configure apn correction mode on AGW,
+          should be one of
+            enable: Enable stateless mode, do nothing if already stateless
+            disable: Disable stateless mode, do nothing if already stateful
```
Comment (ulaskozat) [inline lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py]: fix the comment about stateless feature

**Review by ulaskozat — APPROVED** [review]
LGTM.

**Review by ardzoht — APPROVED** [review]

File: lte/gateway/c/oai/tasks/mme_app/mme_config.c [hunk 2]
```diff
@@ -1054,6 +1064,53 @@ int mme_config_parse_file(mme_config_t* config_pP) {
...
+
+        subsetting = config_setting_get_member(setting,
+            MME_CONFIG_STRING_NAS_APN_CORRECTION_MAP_LIST);
+        config_pP->nas_config.apn_map_config.nb = 0;
+        if (subsetting != NULL) {
+          num = config_setting_length(subsetting);
+          OAILOG_INFO(LOG_MME_APP, "Number of apn correction map configured =%d\n", num);
+          AssertFatal(
+              num <= MAX_APN_CORRECTION_MAP_LIST,
+              "Number of apn correction mmap configured:%d exceeds the maximum number supported"
```
Comment (ssanadhya) [inline lte/gateway/c/oai/tasks/mme_app/mme_config.c]: Typo "mmap"

File: lte/gateway/python/integ_tests/s1aptests/test_attach_esm_information_apn_correction.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+from integ_tests.s1aptests import s1ap_wrapper
+from s1ap_utils import MagmadUtil
+import ctypes
+
+
+class TestEsmInformationWithApnCorrection(unittest.TestCase):
+    def setUp(self):
+        self._s1ap_wrapper = s1ap_wrapper.TestWrapper(
+                stateless_mode=MagmadUtil.stateless_cmds.DISABLE,
```
Comment (ssanadhya) [inline lte/gateway/python/integ_tests/s1aptests/test_attach_esm_information_apn_correction.py]: Stateless mode is disabled by default so no need to specify it here explicitly

File: lte/gateway/python/integ_tests/s1aptests/test_standalone_pdn_conn_req_with_apn_correction.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+import s1ap_types
+import s1ap_wrapper
+import ctypes
+
+
+class TestStandAlonePdnConnReqWithApnCorrection(unittest.TestCase):
+
+    def setUp(self):
+        self._s1ap_wrapper = s1ap_wrapper.TestWrapper(
+                stateless_mode=MagmadUtil.stateless_cmds.DISABLE,
```
Comment (ssanadhya) [inline lte/gateway/python/integ_tests/s1aptests/test_standalone_pdn_conn_req_with_apn_correction.py]: Same comment as above

File: lte/gateway/python/integ_tests/s1aptests/test_standalone_pdn_conn_req_with_apn_correction.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        req.pdnType_pr.pres = 1
+        # PDN Type = IPv4
+        req.pdnType_pr.pdn_type = 1
+        req.pdnAPN_pr.pres = 1
+        s = 'internet.mnc012.mcc345.gprs'
+        req.pdnAPN_pr.len = len(s)
+        req.pdnAPN_pr.pdn_apn = (ctypes.c_ubyte * 100)(*[ctypes.c_ubyte(ord(c)) for c in s[:100]])
+        self._s1ap_wrapper.s1_util.issue_cmd(
+            s1ap_types.tfwCmd.UE_PDN_CONN_REQ, req)
+        # Receive PDN Connectivity Reject
```
Comment (ssanadhya) [inline lte/gateway/python/integ_tests/s1aptests/test_standalone_pdn_conn_req_with_apn_correction.py:75]: Why are we expecting a reject here if APN correction is enabled?

**Comment by themarwhal** [comment]
Unmerging because S1AP tests seems to be consistently failing on CI... Will merge again after jenkins run

**Pull request merged**